generate/sequentially.py

```diff
@@ -0,0 +1,235 @@
+
+import generate.base as generate_base
+from lit_gpt import GPT, Config, Tokenizer
+from lit_gpt.model import Block, build_mask_cache
+from lit_gpt.utils import check_valid_checkpoint_dir, get_default_supported_precision
+
+
+@torch.inference_mode()
+def sequential(model: GPT, root: torch.device, max_seq_length: int, devices: int):
+    if model.config.n_layer % devices:
+        # TODO: support smarter partitioning schemes
+        raise NotImplementedError(
+            f"Only balanced partitioning is implemented: n_layer={model.config.n_layer}, devices {devices}"
+        )
+    layers_per_rank = model.config.n_layer // devices
+    # dictates where each block should be instantiated
+    mapping = layer_to_device(model, chunk_on=Block, chunk_size=layers_per_rank)
+
+    # materialize each block on the appropriate device
+    for path, target_index in mapping.items():
+        submodule = model.get_submodule(path)
+        target_device = torch.device(root.type, target_index)
+        print(f"Moving {path!r} to {target_device}", file=sys.stderr)
+        # submodules loaded by the checkpoint will be on CPU (if no quantization). move them
+        replace_device(submodule, replace=torch.device("cpu"), by=target_device)
+        # in case the checkpoint was partial, materialize leftover metas
+        _materialize_meta_tensors(submodule, target_device)
+        # and build the kv cache
+        submodule.attn.kv_cache = submodule.attn.build_kv_cache(1, max_seq_length, model.cos.size(-1), target_device)
+    # rebuild odd ends
+    with root:
+        # the rope cache which is on meta device
+        model.max_seq_length = max_seq_length
+        # the mask cache which cannot be created with `set_kv_cache` because that will set it for all layers
+        model.mask_cache = build_mask_cache(max_seq_length)
+    # and everything that is not a block in the root
+    _materialize_meta_tensors(model, root)
+    replace_device(model, replace=torch.device("cpu"), by=root)
+
+    if devices > 1:
+        # install hooks to move layer inputs/output between devices
+        for layer_num, (path, target_index) in enumerate(mapping.items()):
+            submodule = model.get_submodule(path)
+            if layer_num >= layers_per_rank:
+                # we need to move the block input on the boundaries between devices
+                # and also on every non-root device because the RoPE and mask cache is shared
+                # TODO: the second case could be optimized and then we would only need this hook for
+                # `layer_num in [layers_per_rank * i - 1 for i in range(1, devices + 1)]`
+                target_device = torch.device(root.type, target_index)
+                submodule.register_forward_pre_hook(partial(move_block_input, target_device))
+            if layer_num == model.config.n_layer - 1:
+                submodule.register_forward_hook(partial(move_block_output, root))
+
+    return model
+
+
+def layer_to_device(
```

Comment:
Shouldn't we do this only on nn.Sequential? The comment in the code already hints at this. The blocks are in nn.Sequential already, so that would be good.

Reply:
The blocks are in a ModuleList, not Sequential. I don't think we can use Sequential because we need to pass `cos, sin, rope, input_pos` in addition to the input